crates/control/src/calibration_controller.rs

```diff
@@ -0,0 +1,273 @@
+                return CalibrationState::Initialize {
+                    started_time: *current_cycle_time,
+                };
+            }
+            CalibrationState::Initialize {
+                started_time: activated_time,
+            } => {
+                let waiting_duration = current_cycle_time
+                    .start_time
+                    .duration_since(activated_time.start_time)
+                    .unwrap_or_default();
+
+                if waiting_duration >= *context.stabilization_delay {
+                    return self.get_next_look_at_or_processing(*context.cycle_time);
+                }
+            }
+            CalibrationState::LookAt {
+                dispatch_time,
+                camera,
+                ..
+            } => {
+                let time_diff = current_cycle_time
+                    .start_time
+                    .duration_since(dispatch_time.start_time)
+                    .unwrap_or_default();
+
+                if time_diff > *context.look_at_dispatch_delay {
+                    self.last_capture_retries = 0;
+                    return CalibrationState::Capture {
+                        camera,
+                        dispatch_time: *current_cycle_time,
+                    };
+                }
+            }
+            CalibrationState::Capture {
+                dispatch_time,
+                camera,
+            } => {
+                let calibration_response = collect_filtered_values(
+                    match camera {
+                        CameraPosition::Top => &context.measurement_top,
+                        CameraPosition::Bottom => &context.measurement_bottom,
+                    },
+                    &dispatch_time,
+                );
+
+                let goto_next_lookat = calibration_response.map_or(false, |response| {
+                    if let Some(measurement) = response.measurement {
+                        self.current_measurements.push(measurement);
+                        true
+                    } else {
+                        self.last_capture_retries += 1;
+                        self.last_capture_retries > *context.max_retries_per_capture
+                    }
+                });
+                if goto_next_lookat {
+                    return self.get_next_look_at_or_processing(*context.cycle_time);
+                }
+            }
+            CalibrationState::Process => {
+                // TODO Handle not enough measurements
+                let solved_result = solve(
+                    Corrections::default(),
+                    self.current_measurements.clone(),
+                    *context.field_dimensions,
+                );
+
+                self.last_calibration_corrections = Some(solved_result);
+                return CalibrationState::Finish;
+            }
+            CalibrationState::Finish => {}
+        }
+
+        return self.current_calibration_state;
```

Comment:
You should decide if this function should change the `self.current_calibration_state` or return the new state.
Currently you do both and then assign the return value to `self.current_calibration_state` again.

Reply:
I noticed the same today xD and changed as follows:
1. Move the big-match into `cycle`, 
2. but take the larger parts into own functions -> i.e. process capture, etc
3. While we eliminated `Process`, I didn't want to do that part in an anonymous function, so a new function called `calibrate` is done.

I still think this can be refactored better, but if we don't have any other serious issues, we could try to finalize this :)

P.S. In case you feel moving the `match` is too much, i can revert that commit.